# Pytest Exercise
 
In this exercise, we'll get hands-on experience using `PyTest`, a software testing tool for Python.

**PyTest** is a popular Python testing framework that makes it easy to write and run tests. 

We'll practice:
 - Testing an SuT and obtaining reports on **statement coverage** and **branch coverage**.

## Exercise Environment

This exercise uses the following file structure:

```
exercises/
├── classify_triangle.py       # System under Test (SuT)
└── test_classify_triangle.py  # Test code for pytest
```

| File | Description |
|------|-------------|
| [`classify_triangle.py`](./classify_triangle.py) | Contains the `classify_triangle` function and a `main` entry point |
| [`test_classify_triangle.py`](./test_classify_triangle.py) | Contains test functions for pytest |

## Running PyTest

`PyTest` automatically discovers test files and functions based on naming conventions.
It runs tests by looking for: 
- Files named `test_*.py` or `*_test.py`
- Functions prefixed with `test_` inside those files

It searches the current directory and all subdirectories.

### In This Exercise

Running `pytest` will execute all three `test_*` functions defined in `test_classify_triangle.py`.

Run the three tests by executing the following command:

In [3]:
!pytest

============================= test session starts ==============================
platform linux -- Python 3.10.19, pytest-8.3.3, pluggy-1.6.0
rootdir: /workspace/modules/02_test_coverages/exercises
plugins: anyio-4.11.0, cov-5.0.0
collected 3 items                                                              

test_classify_triangle.py ...                                            [100%]

============================== 3 passed in 0.01s ===============================


## Measuring Statement and Branch Coverage with PyTest

The plain `pytest` command only runs tests but does not collect coverage data. To measure coverage, we use the `pytest-cov` plugin.

### Statement Coverage

Use the `--cov` option to measure **statement coverage**:

```bash
pytest --cov=<module_name>
```

| Column | Description |
|--------|-------------|
| **Stmts** | Total number of executable statements |
| **Miss** | Number of statements not executed |
| **Cover** | Percentage of statements executed |

Run the following command to measure statement coverage for `classify_triangle.py`:

In [4]:
!pytest --cov=classify_triangle

============================= test session starts ==============================
platform linux -- Python 3.10.19, pytest-8.3.3, pluggy-1.6.0
rootdir: /workspace/modules/02_test_coverages/exercises
plugins: anyio-4.11.0, cov-5.0.0
collected 3 items                                                              

test_classify_triangle.py ...                                            [100%]

---------- coverage: platform linux, python 3.10.19-final-0 ----------
Name                   Stmts   Miss  Cover
------------------------------------------
classify_triangle.py      23      6    74%
------------------------------------------
TOTAL                     23      6    74%


============================== 3 passed in 0.02s ===============================


### Branch Coverage

To also measure **branch coverage** (ensuring every `if`/`else` path is tested), add the `--cov-branch` option:

```bash
pytest --cov=<module_name> --cov-branch
```

This adds two more columns to the output:

| Column | Description |
|--------|-------------|
| **Branch** | Total number of branches in the code |
| **BrPart** | Number of branches only partially covered |
| **Cover** | Combined coverage (%) calculated as:  `((Stmts-Miss) + (Branch-BrPart)) / (Stmts + Branch) * 100` |

Run the following command to measure both statement and branch coverage:

In [5]:
!pytest --cov=classify_triangle --cov-branch

============================= test session starts ==============================
platform linux -- Python 3.10.19, pytest-8.3.3, pluggy-1.6.0
rootdir: /workspace/modules/02_test_coverages/exercises
plugins: anyio-4.11.0, cov-5.0.0
collected 3 items                                                              

test_classify_triangle.py ...                                            [100%]

---------- coverage: platform linux, python 3.10.19-final-0 ----------
Name                   Stmts   Miss Branch BrPart  Cover
--------------------------------------------------------
classify_triangle.py      23      6     14      5    70%
--------------------------------------------------------
TOTAL                     23      6     14      5    70%


============================== 3 passed in 0.02s ===============================


## Generating an HTML Coverage Report

The terminal output provides a quick summary, but for detailed analysis, you can generate an **interactive HTML report**.

```bash
coverage html
```

Run the following command.

In [6]:
!coverage html

Wrote HTML report to ]8;;file:///workspace/modules/02_test_coverages/exercises/htmlcov/index.htmlhtmlcov/index.html]8;;


This creates a `htmlcov/` directory containing:
- `index.html` — Overview of all files with coverage percentages
- Individual file pages — Line-by-line highlighting showing:
  - ✅ **Green**: Executed lines
  - ❌ **Red**: Missed lines
  - 🟡 **Yellow**: Partially covered branches

Open `htmlcov/index.html` in a browser to explore the detailed report.

## 🛠️ Running the `coverage` Tool Directly

So far, we've used `pytest-cov` to measure coverage through test functions. But what if you want to measure coverage by **running a script directly** with specific inputs?

The `coverage` command-line tool allows you to:
- Run any Python script while tracking coverage
- Accumulate coverage data across multiple runs
- Generate reports without writing test functions

### Key Commands

| Command | Description |
|---------|-------------|
| `coverage erase` | Clear previous coverage data |
| `coverage run script.py [args]` | Run a script and collect coverage |
| `coverage run -a script.py [args]` | Append coverage data to existing data |
| `coverage report` | Display coverage summary in terminal |
| `coverage html` | Generate detailed HTML report |

### Statement Coverage

Try running `classify_triangle.py` directly with different inputs and observe how statement coverage accumulates:

In [7]:
# 1) 이전 커버리지 데이터 삭제
!coverage erase

# 2) 여러 입력으로 스크립트 실행 (append 모드 -a)
!coverage run classify_triangle.py 3 3 3
!coverage run -a classify_triangle.py 3 4 5
!coverage run -a classify_triangle.py 4 5 6

# 3) 리포트 보기
!coverage report
!coverage html

equilateral
right
scalene
Name                   Stmts   Miss  Cover
------------------------------------------
classify_triangle.py      23      4    83%
------------------------------------------
TOTAL                     23      4    83%
Wrote HTML report to ]8;;file:///workspace/modules/02_test_coverages/exercises/htmlcov/index.htmlhtmlcov/index.html]8;;


### Branch Coverage

To track branch coverage, add the `--branch` flag. Try running with different inputs and observe how branch coverage accumulates:

In [8]:
# 1) 이전 커버리지 데이터 삭제
!coverage erase

# 2) 여러 입력으로 스크립트 실행 (append 모드 -a, --branch 플래그 추가)
!coverage run --branch classify_triangle.py 3 3 3
!coverage run -a --branch classify_triangle.py 3 4 5
!coverage run -a --branch classify_triangle.py 4 5 6

# 3) 리포트 보기
!coverage report
!coverage html

equilateral
right
scalene
Name                   Stmts   Miss Branch BrPart  Cover
--------------------------------------------------------
classify_triangle.py      23      4     14      5    76%
--------------------------------------------------------
TOTAL                     23      4     14      5    76%
Wrote HTML report to ]8;;file:///workspace/modules/02_test_coverages/exercises/htmlcov/index.htmlhtmlcov/index.html]8;;


Check out the HTML report in the `htmlcov/` directory.

### 🤔 Quiz: Why Are the Coverage Results Different?

You may have noticed that the coverage results differ between `pytest --cov` and `coverage run` with the same test suite:

| Method | Statement Coverage |
|--------|-------------------|
| `pytest --cov` | 74% |
| `coverage run` (direct execution) | 96% |

**Why do you think there is a difference?**

**Hint**: Compare the entry points for each approach.